relay/nakama/siwe/siwe.go

```diff
@@ -0,0 +1,278 @@
+	Version string         `json:"-"`
+}
+
+// nonceWindow represents a single nonce that was used as well as the window of time that it was valid for.
+type nonceWindow struct {
+	Nonce     string    `json:"nonce"`
+	IssuedAt  time.Time `json:"issued_at"`
+	ExpiredAt time.Time `json:"expired_at"`
+}
+
+// ValidateSignature ensures the given message is valid and the given signature is actually a signature for the message.
+func ValidateSignature(ctx context.Context, nk runtime.NakamaModule, signerAddress, message, signature string) error {
+	msg, err := siwe.ParseMessage(message)
+	if err != nil {
+		return eris.Wrap(err, "failed to parse message")
+	}
+
+	if msg.GetDomain() != DefaultDomain {
+		return eris.New("domain is incorrect")
+	}
+	if uri := msg.GetURI(); uri.String() != DefaultURI {
+		return eris.New("uri is incorrect")
+	}
+
+	nonce, err := getNonceWindow(msg)
+	if err != nil {
+		return eris.Wrap(err, "nonce does not have a valid time window")
+	}
+
+	// Make sure the signature is valid.
+	pubKey, err := msg.VerifyEIP191(signature)
+	if err != nil {
+		return eris.Wrap(err, "failed to verify signature")
+	}
+
+	// Make sure the signer of the message and the signer in the message match
+	actualSignerAddress := crypto.PubkeyToAddress(*pubKey).Hex()
+	if actualSignerAddress != signerAddress {
+		return eris.Wrap(err, "signer address in message and signature are different ")
+	}
+
+	// Make sure the nonce is valid and hasn't yet been used.
+	nonceUsed, err := useNonce(ctx, nk, actualSignerAddress, nonce)
+	if err != nil {
+		return eris.Wrap(err, "failed to use nonce")
+	} else if !nonceUsed {
+		return eris.New("nonce has already been used")
+	}
+	return nil
+}
+
+func getNonceWindow(msg *siwe.Message) (*nonceWindow, error) {
+	// Make sure the expiration time is in the future
+	expString := msg.GetExpirationTime()
+	if expString == nil {
+		return nil, eris.New("missing expiration time")
+	}
+	expTime, err := time.Parse(time.RFC3339, *expString)
+	if err != nil {
+		return nil, eris.New("unable to parse expiration time")
+	}
+	if time.Now().After(expTime) {
+		return nil, eris.New("message has expired")
+	}
+
+	// Make sure the issued time is in the past
+	issuedAt, err := time.Parse(time.RFC3339, msg.GetIssuedAt())
+	if err != nil {
+		return nil, eris.New("failed to parse issued at time")
+	}
+	if time.Now().Before(issuedAt) {
+		return nil, eris.New("message must be issued in the past")
+	}
+
+	delta := expTime.Sub(issuedAt)
+	if delta > DefaultTimeout {
+		return nil, eris.New("delta between expiration time and issued at time is too large")
+	}
+	return &nonceWindow{
+		Nonce:     msg.GetNonce(),
+		ExpiredAt: expTime,
+		IssuedAt:  issuedAt,
+	}, nil
+}
+
+// useNonce attempts to use the given nonce for the given signer address. If an identical nonce has been used, and
+// that nonce has not yet expired, useNonce will return false. A return value of (true, nil) means the nonce was
+// successfully used.
+func useNonce(
```

Comment:
a little confused on the multiple nonces stuff. can you explain?

Reply:
I tried to clarify this comment. If it's still not clear, let's jump on a call and talk about what's confusing. Then I'll take another swing and the phrasing.

Reply:
that helps. all good here